server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java

```diff
@@ -193,4 +195,32 @@ public static long loadVersion(IndexReader reader, Term term) throws IOException
         final DocIdAndVersion docIdAndVersion = loadDocIdAndVersion(reader, term);
         return docIdAndVersion == null ? NOT_FOUND : docIdAndVersion.version;
     }
+
+    /**
+     * Checks for the existence of the history of a pair SeqNo/PrimaryTerm in Lucene. The checking pair is considered as existed
+     * if there is a pair such as the seqNo equals to the checking seqNo and the primary term is at least the checking term.
+     */
+    public static boolean hasHistoryInLucene(IndexReader reader, Term idTerm, long seqNo, long primaryTerm) throws IOException {
+        final PerThreadIDVersionAndSeqNoLookup[] lookups = getLookupState(reader, idTerm.field());
+        final List<LeafReaderContext> leaves = reader.leaves();
+        // iterate backwards to optimize for the frequently updated documents which are likely to be in the last segments
+        for (int i = leaves.size() - 1; i >= 0; i--) {
+            final LeafReaderContext leaf = leaves.get(i);
+            final PerThreadIDVersionAndSeqNoLookup lookup = lookups[leaf.ord];
+            final PostingsEnum postingsEnum = lookup.getPostingsOrNull(idTerm.bytes());
+            if (postingsEnum == null) {
+                continue;
+            }
+            final NumericDocValues seqNoDV = leaf.reader().getNumericDocValues(SeqNoFieldMapper.NAME);
```

Comment:
I think we can assert that both seqNoDV and primaryTermDV are non null. They are required at least for this code to work.